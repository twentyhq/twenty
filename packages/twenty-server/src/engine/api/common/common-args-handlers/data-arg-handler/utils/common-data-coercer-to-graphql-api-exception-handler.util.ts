import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';
import { UserInputError } from 'src/engine/core-modules/graphql/utils/graphql-errors.util';

export const commonDataCoercerToGraphqlApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new UserInputError(error);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
