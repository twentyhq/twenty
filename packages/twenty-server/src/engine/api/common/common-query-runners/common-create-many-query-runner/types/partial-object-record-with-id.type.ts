import { type ObjectRecord } from 'twenty-shared/types';

export type PartialObjectRecordWithId = Partial<ObjectRecord> & { id: string };
