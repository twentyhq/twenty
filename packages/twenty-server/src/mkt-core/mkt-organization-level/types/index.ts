export * from './organization-level.types';
export * from './permissions.type';
