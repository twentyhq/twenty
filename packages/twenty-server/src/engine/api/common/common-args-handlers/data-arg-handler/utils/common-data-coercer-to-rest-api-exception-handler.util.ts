import { BadRequestException } from '@nestjs/common';

import { assertUnreachable } from 'twenty-shared/utils';

import {
  type CommonDataCoercerException,
  CommonDataCoercerExceptionCode,
} from 'src/engine/api/common/common-args-handlers/data-arg-handler/errors/common-data-coercer.exception';

export const commonDataCoercerToRestApiExceptionHandler = (
  error: CommonDataCoercerException,
) => {
  switch (error.code) {
    case CommonDataCoercerExceptionCode.INVALID_COMPOSITE_FIELD:
    case CommonDataCoercerExceptionCode.INVALID_NUMBER:
    case CommonDataCoercerExceptionCode.INVALID_TEXT:
    case CommonDataCoercerExceptionCode.INVALID_DATE_OR_DATE_TIME:
    case CommonDataCoercerExceptionCode.INVALID_BOOLEAN:
    case CommonDataCoercerExceptionCode.INVALID_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_MULTI_SELECT:
    case CommonDataCoercerExceptionCode.INVALID_UUID:
    case CommonDataCoercerExceptionCode.INVALID_ARRAY:
    case CommonDataCoercerExceptionCode.INVALID_RAW_JSON:
      throw new BadRequestException(error.message);
    default: {
      return assertUnreachable(error.code);
    }
  }
};
