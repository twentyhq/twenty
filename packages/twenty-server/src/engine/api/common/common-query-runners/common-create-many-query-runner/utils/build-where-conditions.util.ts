import { isDefined } from 'twenty-shared/utils';
import { type FindOperator, In } from 'typeorm';

import { type ObjectRecord } from 'src/engine/api/graphql/workspace-query-builder/interfaces/object-record.interface';

import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';

export const buildWhereConditions = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
): Record<string, FindOperator<string>>[] => {
  const whereConditions = [];

  for (const field of conflictingFields) {
    const fieldValues = records
      .map((record) => getValueFromPath(record, field.fullPath))
      .filter(isDefined);

    if (fieldValues.length > 0) {
      whereConditions.push({ [field.column]: In(fieldValues) });
    }
  }

  return whereConditions;
};
