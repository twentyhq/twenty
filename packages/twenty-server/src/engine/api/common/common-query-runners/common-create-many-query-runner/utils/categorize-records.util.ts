import { isDefined } from 'twenty-shared/utils';

import { type ObjectRecord } from 'twenty-shared/types';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getMatchingRecordId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-matching-record-id.util';

export const categorizeRecords = (
  records: Partial<ObjectRecord>[],
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): {
  recordsToUpdate: PartialObjectRecordWithId[];
  recordsToInsert: Partial<ObjectRecord>[];
} => {
  const recordsToUpdate: PartialObjectRecordWithId[] = [];
  const recordsToInsert: Partial<ObjectRecord>[] = [];

  for (const record of records) {
    const matchingRecordId = getMatchingRecordId(
      record,
      conflictingFields,
      existingRecords,
    );

    if (isDefined(matchingRecordId)) {
      recordsToUpdate.push({ ...record, id: matchingRecordId });
    } else {
      recordsToInsert.push(record);
    }
  }

  return { recordsToUpdate, recordsToInsert };
};
