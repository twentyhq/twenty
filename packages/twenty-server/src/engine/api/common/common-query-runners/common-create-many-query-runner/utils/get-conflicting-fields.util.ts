import { capitalize } from 'twenty-shared/utils';

import { compositeTypeDefinitions } from 'src/engine/metadata-modules/field-metadata/composite-types';
import { type ObjectMetadataItemWithFieldMaps } from 'src/engine/metadata-modules/types/object-metadata-item-with-field-maps';

export const getConflictingFields = (
  objectMetadataItemWithFieldMaps: ObjectMetadataItemWithFieldMaps,
): {
  baseField: string;
  fullPath: string;
  column: string;
}[] => {
  return Object.values(objectMetadataItemWithFieldMaps.fieldsById)
    .filter((field) => field.isUnique || field.name === 'id')
    .flatMap((field) => {
      const compositeType = compositeTypeDefinitions.get(field.type);

      if (!compositeType) {
        return [
          {
            baseField: field.name,
            fullPath: field.name,
            column: field.name,
          },
        ];
      }

      const property = compositeType.properties.find(
        (prop) => prop.isIncludedInUniqueConstraint,
      );

      return property
        ? [
            {
              baseField: field.name,
              fullPath: `${field.name}.${property.name}`,
              column: `${field.name}${capitalize(property.name)}`,
            },
          ]
        : [];
    });
};
