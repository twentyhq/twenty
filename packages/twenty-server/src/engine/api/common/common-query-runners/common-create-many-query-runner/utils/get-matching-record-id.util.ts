import { msg } from '@lingui/core/macro';
import { type ObjectRecord } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';

import { type PartialObjectRecordWithId } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/types/partial-object-record-with-id.type';
import { getValueFromPath } from 'src/engine/api/common/common-query-runners/common-create-many-query-runner/utils/get-value-from-path.util';
import {
  CommonQueryRunnerException,
  CommonQueryRunnerExceptionCode,
} from 'src/engine/api/common/common-query-runners/errors/common-query-runner.exception';

export const getMatchingRecordId = (
  record: Partial<ObjectRecord>,
  conflictingFields: {
    baseField: string;
    fullPath: string;
    column: string;
  }[],
  existingRecords: PartialObjectRecordWithId[],
): string | undefined => {
  const matchingRecordIds = conflictingFields.reduce<string[]>((acc, field) => {
    const requestFieldValue = getValueFromPath(record, field.fullPath);

    const matchingRecord = existingRecords.find((existingRecord) => {
      const existingFieldValue = getValueFromPath(
        existingRecord,
        field.fullPath,
      );

      return (
        isDefined(existingFieldValue) &&
        existingFieldValue === requestFieldValue
      );
    });

    if (isDefined(matchingRecord)) {
      acc.push(matchingRecord.id);
    }

    return acc;
  }, []);

  if ([...new Set(matchingRecordIds)].length > 1) {
    const conflictingFieldsValues = conflictingFields
      .map((field) => {
        const value = getValueFromPath(record, field.fullPath);

        return isDefined(value) ? `${field.fullPath}: ${value}` : undefined;
      })
      .filter(isDefined)
      .join(', ');

    throw new CommonQueryRunnerException(
      `Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      CommonQueryRunnerExceptionCode.UPSERT_MULTIPLE_MATCHING_RECORDS_CONFLICT,
      {
        userFriendlyMessage: msg`Multiple records found with the same unique field values for ${conflictingFieldsValues}. Cannot determine which record to update.`,
      },
    );
  }

  return matchingRecordIds[0];
};
