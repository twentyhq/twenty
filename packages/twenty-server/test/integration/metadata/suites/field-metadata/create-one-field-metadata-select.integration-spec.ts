import {
  UPDATE_CREATE_ONE_FIELD_METADATA_SELECT_TEST_CASES,
  UpdateCreateFieldMetadataSelectTestCase,
} from 'test/integration/metadata/suites/field-metadata/update-create-one-field-metadata-select-tests-cases';
import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import {
  LISTING_NAME_PLURAL,
  LISTING_NAME_SINGULAR,
} from 'test/integration/metadata/suites/object-metadata/constants/test-object-names.constant';
import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { FieldMetadataType } from 'twenty-shared/types';
import { isDefined } from 'twenty-shared/utils';
import { createOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata.util';

import { FieldMetadataComplexOption } from 'src/engine/metadata-modules/field-metadata/dtos/options.input';

const { failingTestCases, successfulTestCases } =
  UPDATE_CREATE_ONE_FIELD_METADATA_SELECT_TEST_CASES;

describe('Field metadata select creation tests group', () => {
  let createdObjectMetadataId: string;

  beforeEach(async () => {
    const { data } = await createOneObjectMetadata({
      input: {
        labelSingular: LISTING_NAME_SINGULAR,
        labelPlural: LISTING_NAME_PLURAL,
        nameSingular: LISTING_NAME_SINGULAR,
        namePlural: LISTING_NAME_PLURAL,
        icon: 'IconBuildingSkyscraper',
        isLabelSyncedWithName: false,
      },
    });

    createdObjectMetadataId = data.createOneObject.id;
  });

  afterEach(async () => {
    await deleteOneObjectMetadata({
      input: { idToDelete: createdObjectMetadataId },
    });
  });

  test.each(successfulTestCases)(
    'Create $title',
    async ({ context: { input, expectedOptions } }) => {
      const { data, errors } = await createOneFieldMetadata({
        input: {
          objectMetadataId: createdObjectMetadataId,
          type: FieldMetadataType.SELECT,
          name: 'testField',
          label: 'Test Field',
          isLabelSyncedWithName: false,
          ...input,
        },
        gqlFields: `
        id
        options
        defaultValue
        `,
      });

      expect(data).not.toBeNull();
      expect(data.createOneField).toBeDefined();
      const createdOptions: FieldMetadataComplexOption[] =
        data.createOneField.options;

      const optionsToCompare = expectedOptions ?? input.options;

      expect(errors).toBeUndefined();
      expect(createdOptions.length).toBe(optionsToCompare.length);
      createdOptions.forEach((option) => expect(option.id).toBeDefined());
      expect(createdOptions).toMatchObject(optionsToCompare);

      if (isDefined(input.defaultValue)) {
        expect(data.createOneField.defaultValue).toEqual(input.defaultValue);
      }
    },
  );

  const createSpecificFailingTestCases: UpdateCreateFieldMetadataSelectTestCase[] =
    [
      {
        title: 'should fail with null options',
        context: {
          input: {
            options: null as unknown as FieldMetadataComplexOption[],
          },
        },
      },
      {
        title: 'should fail with undefined options',
        context: {
          input: {
            options: undefined as unknown as FieldMetadataComplexOption[],
          },
        },
      },
    ];

  test.each([...failingTestCases, ...createSpecificFailingTestCases])(
    'Create $title',
    async ({ context: { input } }) => {
      const { data, errors } = await createOneFieldMetadata({
        input: {
          objectMetadataId: createdObjectMetadataId,
          type: FieldMetadataType.SELECT,
          name: 'testField',
          label: 'Test Field',
          isLabelSyncedWithName: false,
          ...input,
        },
        gqlFields: `
        id
        options
        `,
      });

      expect(data).toBeNull();
      expect(errors).toBeDefined();
      expect(errors).toMatchSnapshot();
    },
  );
});
