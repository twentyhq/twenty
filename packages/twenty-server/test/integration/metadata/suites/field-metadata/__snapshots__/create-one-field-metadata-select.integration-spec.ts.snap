// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Field metadata select creation tests group Create should fail with an invalid default value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Default value for existing options is invalid: invalid",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with an unknown default value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Default value for existing options is invalid: 'OPTION_424242'",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with comma in option label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Label must not contain a comma",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with duplicated option ids 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option id "fd1f11fd-3f05-4a33-bddf-800c3412ce98"",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with duplicated option positions 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option position "1"",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with duplicated option values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option value "OPTION_1"",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with duplicated trimmed option values 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Duplicated option value "OPTION_1"",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with empty options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Options are required for enum fields",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with empty string id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with empty string label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with empty string value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with invalid option id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with invalid option value format 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "Option 1 and some other things, /"",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with not a string id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with not a string label 1`] = `
[
  {
    "extensions": {
      "code": "INTERNAL_SERVER_ERROR",
      "exceptionEventId": "mocked-exception-id",
    },
    "message": "label.includes is not a function",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with not a string value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Value must be in UPPER_CASE and follow snake_case "22222"",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with null id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is required",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with null label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label is required",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with null options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Options are required for enum fields",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with null value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value is required",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with only white spaces id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with only white spaces label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with only white spaces value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value "" is beneath 1 character",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with too long id 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option id is invalid",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with too long label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" exceeds 63 characters",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with too long value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa" exceeds 63 characters",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with undefined option label 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option label is required",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with undefined option value 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Option value is required",
  },
]
`;

exports[`Field metadata select creation tests group Create should fail with undefined options 1`] = `
[
  {
    "extensions": {
      "code": "BAD_USER_INPUT",
    },
    "message": "Options are required for enum fields",
  },
]
`;
